### PR TITLE
[Feat] Support zero bubble with shardformer input

### LINKED ISSUE #6037: [FEATURE]: Support Zerobubble pipeline
### Describe the feature

This issue contains requirements for the zerobubble pp.
Source Github Link: https://github.com/sail-sg/zero-bubble-pipeline-parallelism
Source Paper: https://arxiv.org/abs/2401.10241

### PR DESCRIPTION
**Pull request by duanjunwen**
## 📌 Checklist before creating the PR

- [x] I have created an issue for this PR for traceability
- [x] The title follows the standard format: `[doc/gemini/tensor/...]: A concise description`
- [x] I have added relevant tags if possible for us to better distinguish different PRs
- [x] I have installed pre-commit: `pip install pre-commit && pre-commit install`

## 🚨 Issue number

> Link this PR to your issue with words like fixed to automatically close the linked issue upon merge
>
> e.g. `fixed #1234`, `closed #1234`, `resolved #1234`
resolved #6037

## 📝 What does this PR do?

> Summarize your work here.
> if you have any plots/diagrams/screenshots/tables, please attach them here.

## 💥 Checklist before requesting a review

- [x] I have linked my PR to an issue ([instruction](https://docs.github.com/en/issues/tracking-your-work-with-issues/linking-a-pull-request-to-an-issue))
- [x] My issue clearly describes the problem/feature/proposal, with diagrams/charts/table/code if possible
- [x] I have performed a self-review of my code
- [x] I have added thorough tests.
- [ ] I have added docstrings for all the functions/methods I implemented

## ⭐️ Do you enjoy contributing to Colossal-AI?

- [x] 🌝 Yes, I do.
- [ ] 🌚 No, I don't.

Tell us more if you don't enjoy contributing to Colossal-AI.

### CODE REVIEW

File: colossalai/pipeline/schedule/zero_bubble_pp.py [hunk 1]
```diff
@@ -39,6 +39,20 @@ def deallocate_output_tensor(out, deallocate_pipeline_outputs=False):
...
 
 
+def require_grad(tensor):
+    """Pseudo-deallocate (i.e., set to scalar) the output tensor's '.data' field.
+
+    This method should be called right after the output tensor has been
+    sent to the next pipeline stage. At this point, the output tensor is
+    only useful for its '.grad_fn' field, and not its '.data'.
+    """
+    if tensor is None:
```
Comment (ver217) [inline colossalai/pipeline/schedule/zero_bubble_pp.py:49]: `if not isinstance(tensor, torch.Tensor):`

File: colossalai/pipeline/schedule/_utils.py [hunk 2]
```diff
@@ -131,6 +131,16 @@ def retain_grad(x: Any) -> None:
         x.retain_grad()
 
 
+def require_grad(x: Any) -> None:
+    """Call require_grad on a tensor.
+
+    Args:
+        x (Any): Object to be called.
+    """
+    if isinstance(x, torch.Tensor) and x.requires_grad:
```
Comment (ver217) [inline colossalai/pipeline/schedule/_utils.py:140]: `and not x.requires_grad`

File: colossalai/pipeline/schedule/zero_bubble_pp.py (hunk 1 above)
Comment (ver217) [reply colossalai/pipeline/schedule/zero_bubble_pp.py:49]: and this function is redundant, which is defined in `_utils.py`

File: colossalai/pipeline/schedule/zero_bubble_pp.py [hunk 3]
```diff
@@ -427,18 +442,27 @@ def forward_step(
...
+                if input_obj is None:
+                    output_obj = model_chunk[model_chunk_id](**micro_batch)
+                else:
+                    output_obj = model_chunk[model_chunk_id](**input_obj)
+            else:
+                # fwd for shardformer
+                # NOTE: in shardformer, each device still has the entire model, so we need to use relevant stage layers
+                internal_inputs = {} if input_obj is None else input_obj
+                # internal_inputs["stage_index"] = self.stage_manager.stage_indices[model_chunk_id]
+                output_obj = model_forward(model_chunk[model_chunk_id], micro_batch, internal_inputs)
```
Comment (ver217) [inline colossalai/pipeline/schedule/zero_bubble_pp.py:461]: should always execute this branch

File: colossalai/pipeline/schedule/zero_bubble_pp.py [hunk 4]
```diff
@@ -472,19 +496,25 @@ def backward_b_step(
...
+            tree_map(retain_grad, input_obj)
+        input_obj_ = input_obj["hidden_states"]
 
         if model_chunk_id == 1 and self.stage_manager.is_first_stage(ignore_chunk=True):
             # loss backward; output_obj is loss; so output_obj_grad should be None
             assert output_obj_grad is None
-
+            output_obj_ = output_obj
+        else:
+            output_obj_ = output_obj["hidden_states"]
```
Comment (ver217) [inline colossalai/pipeline/schedule/zero_bubble_pp.py:510]: remove hardcode

File: colossalai/pipeline/schedule/zero_bubble_pp.py [hunk 5]
```diff
@@ -472,19 +496,25 @@ def backward_b_step(
         # calculate bwd b step ; only dx = w*dy;
 
         # Retain the grad on the input_obj.
-        tree_map(retain_grad, input_obj)
+        if input_obj is None:
+            return None
+        else:
+            tree_map(retain_grad, input_obj)
+        input_obj_ = input_obj["hidden_states"]
```
Comment (ver217) [inline colossalai/pipeline/schedule/zero_bubble_pp.py:503]: remove hardcode

File: colossalai/pipeline/schedule/zero_bubble_pp.py (hunk 1 above)
Comment (duanjunwen) [reply colossalai/pipeline/schedule/zero_bubble_pp.py]: Fixed in https://github.com/hpcaitech/ColossalAI/pull/6065/commits/6ee9584b9a2310bdd556ef32c9901828c2aec04d; Also remove redundant function ;

File: colossalai/pipeline/schedule/zero_bubble_pp.py (hunk 3 above)
Comment (duanjunwen) [reply colossalai/pipeline/schedule/zero_bubble_pp.py:446]: > should always execute this branch

Fixed in https://github.com/hpcaitech/ColossalAI/pull/6065/commits/6ee9584b9a2310bdd556ef32c9901828c2aec04d;

File: colossalai/pipeline/schedule/_utils.py (hunk 2 above)
Comment (duanjunwen) [reply colossalai/pipeline/schedule/_utils.py]: > `and not x.requires_grad`

Fixed in https://github.com/hpcaitech/ColossalAI/pull/6065/commits/6ee9584b9a2310bdd556ef32c9901828c2aec04d;

File: colossalai/pipeline/schedule/zero_bubble_pp.py (hunk 1 above)
Comment (duanjunwen) [reply colossalai/pipeline/schedule/zero_bubble_pp.py]: > and this function is redundant, which is defined in `_utils.py`

Fixed in https://github.com/hpcaitech/ColossalAI/pull/6065/commits/6ee9584b9a2310bdd556ef32c9901828c2aec04d; Also remove redundant function ;

File: colossalai/pipeline/schedule/zero_bubble_pp.py (hunk 4 above)
Comment (duanjunwen) [reply colossalai/pipeline/schedule/zero_bubble_pp.py]: > remove hardcode

Check tensor.require grad in input_obj; if True, append output_obj[key] & output_grad_obj[key] to list, else pass;

Fixed in https://github.com/hpcaitech/ColossalAI/pull/6065/commits/349272c71fa9d30c404ca29b394d7e79bfcd2fd0;

File: colossalai/pipeline/schedule/zero_bubble_pp.py (hunk 5 above)
Comment (duanjunwen) [reply colossalai/pipeline/schedule/zero_bubble_pp.py]: > remove hardcode
Fixed in 
https://github.com/hpcaitech/ColossalAI/pull/6065/commits/349272c71fa9d30c404ca29b394d7e79bfcd2fd0;

File: colossalai/pipeline/schedule/zero_bubble_pp.py [hunk 6]
```diff
@@ -472,19 +478,47 @@ def backward_b_step(
...
+        else:
+            tree_map(retain_grad, input_obj)
+
+        # x, y, dy list for backward_by_grad; Type: list[tensor];
+        input_obj_ = []
+        output_obj_ = []
+        output_obj_grad_ = []
+
+        # get x from input_obj to input_obj_
+        for k, v in input_obj.items():
```
Comment (ver217) [inline colossalai/pipeline/schedule/zero_bubble_pp.py:492]: should use tree_map or tree_flatten

File: colossalai/pipeline/schedule/zero_bubble_pp.py [hunk 7]
```diff
@@ -472,19 +478,47 @@ def backward_b_step(
...
             # loss backward; output_obj is loss; so output_obj_grad should be None
             assert output_obj_grad is None
+            output_obj_grad_.append(output_obj_grad)  # None
+            output_obj_.append(output_obj)  # LOSS
+
+        else:
+            for k, v in input_obj.items():
+                if v.requires_grad:
+                    output_obj_.append(output_obj[k])
+                    output_obj_grad_.append(output_obj_grad[k])
```
Comment (ver217) [inline colossalai/pipeline/schedule/zero_bubble_pp.py:506]: should use tree_map or tree_flatten

File: colossalai/pipeline/schedule/zero_bubble_pp.py [hunk 8]
```diff
@@ -472,19 +478,47 @@ def backward_b_step(
...
             retain_graph=True,
         )
-        return input_obj.grad
+
+        # format output_obj_grad
+        if input_obj is not None:
+            input_obj_grad = {}
+            for k, v in input_obj.items():
+                if isinstance(v, torch.Tensor) and v.grad is not None:
+                    input_obj_grad[k] = v.grad
```
Comment (ver217) [inline colossalai/pipeline/schedule/zero_bubble_pp.py:520]: should use tree_map or tree_flatten

File: colossalai/pipeline/schedule/zero_bubble_pp.py (hunk 3 above)
Comment (duanjunwen) [reply colossalai/pipeline/schedule/zero_bubble_pp.py]: Fixed in https://github.com/hpcaitech/ColossalAI/pull/6065/commits/1739df423c79b0c52ff5957b7992c14081d5dd24;
Now take both microbatch and internal_input as input;

File: colossalai/pipeline/schedule/zero_bubble_pp.py (hunk 8 above)
Comment (duanjunwen) [reply colossalai/pipeline/schedule/zero_bubble_pp.py]: > should use tree_map or tree_flatten

Fixed in https://github.com/hpcaitech/ColossalAI/pull/6065/commits/c6d6ee39bda6d0e3aa0a6233796a0d1059eb30dc;

File: colossalai/pipeline/schedule/zero_bubble_pp.py (hunk 7 above)
Comment (duanjunwen) [reply colossalai/pipeline/schedule/zero_bubble_pp.py]: > should use tree_map or tree_flatten

Fixed in https://github.com/hpcaitech/ColossalAI/pull/6065/commits/c6d6ee39bda6d0e3aa0a6233796a0d1059eb30dc;

File: colossalai/pipeline/schedule/zero_bubble_pp.py (hunk 6 above)
Comment (duanjunwen) [reply colossalai/pipeline/schedule/zero_bubble_pp.py]: > should use tree_map or tree_flatten

Fixed in https://github.com/hpcaitech/ColossalAI/pull/6065/commits/c6d6ee39bda6d0e3aa0a6233796a0d1059eb30dc;

File: colossalai/pipeline/schedule/zero_bubble_pp.py [hunk 9]
```diff
@@ -471,20 +469,52 @@ def backward_b_step(
...
+            input_obj_, _ = tree_flatten(micro_batch)
+            output_obj_, _ = tree_flatten(output_obj)  # y
+            output_obj_grad_, _ = tree_flatten(output_obj_grad)  # dy
+
+        # For loss backward; output_obj is loss; output_obj_grad should be None
+        elif model_chunk_id == 1 and self.stage_manager.is_first_stage(ignore_chunk=True):
             assert output_obj_grad is None
+            input_obj_, _ = tree_flatten(input_obj)
+            output_obj_, _ = tree_flatten(output_obj)  # LOSS
+            output_obj_grad_, _ = tree_flatten(output_obj_grad)  # None
```
Comment (ver217) [inline colossalai/pipeline/schedule/zero_bubble_pp.py:492]: If these two variables are known type, can we remove tree_flatten?

File: colossalai/pipeline/schedule/zero_bubble_pp.py [hunk 10]
```diff
@@ -557,55 +596,73 @@ def schedule_f(
...
         )
+
+        # Step3: deallocate output for bwd b & w; (do not detach output)
+        deallocate_output_obj = tree_map(clone, output_obj)
+        if model_chunk_id == 1 and self.stage_manager.is_first_stage(ignore_chunk=True):
+            # We should not deallocate bwd LOSS
+            pass
+        else:
+            # deallocate output
+            tree_map(deallocate, deallocate_output_obj)
```
Comment (ver217) [inline colossalai/pipeline/schedule/zero_bubble_pp.py:624]: Is this wrong? We clone outputs and then deallocate. What do we get?

File: colossalai/pipeline/schedule/zero_bubble_pp.py (hunk 9 above)
Comment (duanjunwen) [reply colossalai/pipeline/schedule/zero_bubble_pp.py:492]: > If these two variables are known type, can we remove tree_flatten?

Of course. in this branch output_obj is Loss (torch.Tensor Type) and output_obj_grad_ is None;

File: colossalai/pipeline/schedule/zero_bubble_pp.py (hunk 10 above)
Comment (duanjunwen) [reply colossalai/pipeline/schedule/zero_bubble_pp.py:624]: > Is this wrong? We clone outputs and then deallocate. What do we get?

I made a bad name for it, "deallocate" function is actually to release tensor.data and keep grad_fn;

File: colossalai/pipeline/schedule/zero_bubble_pp.py [hunk 11]
```diff
@@ -427,18 +424,18 @@ def forward_step(
...
+        # for the first stage, input_obj is None; So,we use micro_batch as input_obj
         # for other stages, input_obj is the output of the previous/next stage containing hidden_states etc.
         # Only attention_mask from micro_batch is used
-
         with self.stage_manager.switch_model_chunk_id(model_chunk_id):
-            # fwd calculate
-            output_obj = model_chunk[model_chunk_id](input_obj)
+            #  fwd calculate
+            internal_inputs = {} if input_obj is None else input_obj
+            # internal_inputs["stage_index"] = self.stage_manager.stage_indices[model_chunk_id]
```
Comment (ver217) [inline colossalai/pipeline/schedule/zero_bubble_pp.py:433]: uncomment this line

File: colossalai/pipeline/schedule/zero_bubble_pp.py [hunk 12]
```diff
@@ -557,55 +596,75 @@ def schedule_f(
...
+        # 3-1:detach output; detach output for send fwd;
         if model_chunk_id == 1 and self.stage_manager.is_first_stage(ignore_chunk=True):
             # We should not detach bwd LOSS
             pass
         else:
-            detached_output_obj = output_obj.clone().detach()
+            # detach output
+            detached_output_obj = tree_map(detach, output_obj)
+        # 3-2 clone output
+        output_obj = tree_map(clone, output_obj)
```
Comment (ver217) [inline colossalai/pipeline/schedule/zero_bubble_pp.py:626]: we should clone detached output obj

Comment (duanjunwen) [reply colossalai/pipeline/schedule/zero_bubble_pp.py]: Fixed in https://github.com/hpcaitech/ColossalAI/pull/6065/commits/7e6f793c5182d7da95e443967be0a6c9777bd01e;

**Review by ver217 — APPROVED** [review]
